<script src="Homework_2_files/header-attrs-2.8/header-attrs.js"></script>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<div id="goals" class="section level2">
<h2>Goals</h2>
<ol style="list-style-type: decimal">
<li>Import counts from mouse E18 brain scATACseq (5K) (provided by 10X Genomics) in R and process them</li>
</ol>
<blockquote>
<p>Processed data for mouse E18 brain scATACseq can be obtained directly from <a href="https://support.10xgenomics.com/single-cell-atac/datasets/1.2.0/atac_v1_E18_brain_fresh_5k">10X Genomics</a></p>
</blockquote>
<ol start="2" style="list-style-type: decimal">
<li><p>Compare <code>Signac</code> processing and <code>SingleCellExperiment</code> processing</p></li>
<li><p>Transfer annotations from mouse E18 brain scRNAseq (5K) to mouse E18 brain scATACseq</p></li>
</ol>
<blockquote>
<p>Processed data for mouse E18 brain scRNAseq can be obtained directly from <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K">10X Genomics</a></p>
</blockquote>
<ol start="4" style="list-style-type: decimal">
<li>Find genes associated to DA peaks and run GO over-enrichment / gene set enrichment analysis</li>
</ol>
</div>
<div id="process-mouse-e18-brain-5k-scatacseq-data" class="section level2">
<h2>1. Process mouse E18 brain 5K scATACseq data</h2>
<div id="download-data" class="section level3">
<h3>Download data</h3>
<p>scATACseq for mouse E18 5K brain can be downloaded from
<a href="https://support.10xgenomics.com/single-cell-atac/datasets/1.2.0/atac_v1_E18_brain_fresh_5k">10X Genomics</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="fu">mkdir</span> -p data/Homework_2/scATAC</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_analysis.tar.gz -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_analysis.tar.gz</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_filtered_peak_bc_matrix.h5 -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_filtered_peak_bc_matrix.h5</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_filtered_tf_bc_matrix.h5 -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_filtered_tf_bc_matrix.h5</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz.tbi -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz.tbi</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_peak_annotation.tsv -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_peak_annotation.tsv</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_peaks.bed -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_peaks.bed</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a><span class="ex">curl</span> https://cg.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_possorted_bam.bam -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_possorted_bam.bam</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_possorted_bam.bam.bai -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_possorted_bam.bam.bai</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_singlecell.csv -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_singlecell.csv</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_web_summary.html -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_web_summary.html</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_cloupe.cloupe -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_cloupe.cloupe</span></code></pre></div>
</div>
<div id="import-data-in-r" class="section level3">
<h3>Import data in R</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="kw">library</span>(Signac)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a><span class="kw">library</span>(BiocParallel)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true"></a><span class="kw">library</span>(plyranges)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true"></a><span class="co">## -- Read brain ATAC-seq counts</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true"></a>brain_assay &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">CreateChromatinAssay</span>(</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true"></a>    <span class="dt">counts =</span> Seurat<span class="op">::</span><span class="kw">Read10X_h5</span>(<span class="st">&quot;data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_filtered_peak_bc_matrix.h5&quot;</span>),</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true"></a>    <span class="dt">fragments =</span> <span class="st">&#39;data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz&#39;</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true"></a>    <span class="dt">sep =</span> <span class="kw">c</span>(<span class="st">&quot;:&quot;</span>, <span class="st">&quot;-&quot;</span>),</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true"></a>    <span class="dt">genome =</span> <span class="st">&quot;mm10&quot;</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true"></a>    <span class="dt">min.cells =</span> <span class="dv">1</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true"></a><span class="co">## -- Read metadata</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true"></a>metadata &lt;-<span class="st"> </span>vroom<span class="op">::</span><span class="kw">vroom</span>(</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true"></a>    <span class="dt">file =</span> <span class="st">&quot;data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_singlecell.csv&quot;</span>,</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true"></a>    <span class="dt">col_names =</span> <span class="ot">TRUE</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true"></a>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(barcode <span class="op">!=</span><span class="st"> &#39;NO_BARCODE&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">cell =</span> barcode) <span class="op">%&gt;%</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(cell <span class="op">%in%</span><span class="st"> </span><span class="kw">colnames</span>(brain_assay)) <span class="op">%&gt;%</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true"></a><span class="st">    </span><span class="kw">column_to_rownames</span>(<span class="st">&#39;barcode&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true"></a><span class="st">    </span><span class="kw">as.data.frame</span>() </span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true"></a></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true"></a><span class="co">## -- Create Seurat object</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true"></a>    <span class="dt">counts =</span> brain_assay,</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true"></a>    <span class="dt">assay =</span> <span class="st">&#39;peaks&#39;</span>,</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true"></a>    <span class="dt">project =</span> <span class="st">&#39;ATAC&#39;</span>,</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true"></a>    <span class="dt">meta.data =</span> metadata</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true"></a>)</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true"></a></span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true"></a><span class="co">## -- Add genome annotations</span></span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true"></a><span class="kw">AnnotationHubquery</span>(AnnotationHub<span class="op">::</span><span class="kw">AnnotationHub</span>(), <span class="kw">c</span>(<span class="st">&#39;annotations&#39;</span>, <span class="st">&#39;gencode&#39;</span>, <span class="st">&#39;GRCm38&#39;</span>))</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true"></a><span class="kw">Annotation</span>(brain) &lt;-<span class="st"> </span>AnnotationHub<span class="op">::</span><span class="kw">AnnotationHub</span>()[[<span class="st">&#39;AH49547&#39;</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(gene_type <span class="op">==</span><span class="st"> &#39;protein_coding&#39;</span>, type <span class="op">==</span><span class="st"> &#39;gene&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">gene_biotype =</span> gene_type)</span></code></pre></div>
</div>
<div id="qcing-data" class="section level3">
<h3>QCing data</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="co">## -- Nucleosome signal </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">NucleosomeSignal</span>(<span class="dt">object =</span> brain)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a>brain<span class="op">$</span>nucleosome_group &lt;-<span class="st"> </span><span class="kw">ifelse</span>(brain<span class="op">$</span>nucleosome_signal <span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span>, <span class="st">&#39;NS &gt; 4&#39;</span>, <span class="st">&#39;NS &lt; 4&#39;</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a><span class="kw">range</span>(brain<span class="op">$</span>nucleosome_signal)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true"></a><span class="kw">table</span>(brain<span class="op">$</span>nucleosome_group)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true"></a><span class="co">## -- Frag. size distribution</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true"></a>frags &lt;-<span class="st"> </span>vroom<span class="op">::</span><span class="kw">vroom</span>(<span class="st">&#39;/data/20210804_scATACseq-workshop/data/MouseBrain/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz&#39;</span>, <span class="dt">col_names =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">setNames</span>(<span class="kw">c</span>(<span class="st">&#39;chr&#39;</span>, <span class="st">&#39;start&#39;</span>, <span class="st">&#39;stop&#39;</span>, <span class="st">&#39;cell&#39;</span>, <span class="st">&#39;cluster&#39;</span>))</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true"></a>p &lt;-<span class="kw">slice_head</span>(frags, <span class="dt">n =</span> <span class="dv">1000000</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">width =</span> stop <span class="op">-</span><span class="st"> </span>start) <span class="op">%&gt;%</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> width)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_bar</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true"></a><span class="st">    </span><span class="kw">xlim</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">800</span>))</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true"></a></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true"></a><span class="co">## -- TSS enrichment</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">TSSEnrichment</span>(brain, <span class="dt">fast =</span> <span class="ot">TRUE</span>)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true"></a>brain<span class="op">$</span>high.tss &lt;-<span class="st"> </span><span class="kw">ifelse</span>(brain<span class="op">$</span>TSS.enrichment <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>, <span class="st">&#39;High&#39;</span>, <span class="st">&#39;Low&#39;</span>)</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true"></a>df &lt;-<span class="st"> </span>frags_chr12 <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true"></a><span class="st">    </span><span class="kw">left_join</span>(dplyr<span class="op">::</span><span class="kw">select</span>(brain<span class="op">@</span>meta.data, cell, high.tss)) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># For each fragment, recover cell TSS enrich. status</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw">rename</span>(<span class="kw">c</span>(<span class="st">&#39;seqnames&#39;</span> =<span class="st"> &#39;chr&#39;</span>, <span class="st">&#39;end&#39;</span> =<span class="st"> &#39;stop&#39;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_granges</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true"></a><span class="st">    </span><span class="kw">join_overlap_left</span>(plyranges<span class="op">::</span><span class="kw">select</span>(mm_promoters, prom_id)) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Link each fragment to overlapping promoter</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true"></a><span class="st">    </span><span class="kw">drop_na</span>(prom_id) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Remove fragments which are not overlapping any promoter</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true"></a><span class="st">    </span><span class="kw">left_join</span>(<span class="kw">as_tibble</span>(mm_promoters), <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&#39;prom_id&#39;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true"></a>        <span class="dt">midprom =</span> start.y <span class="op">+</span><span class="st"> </span>(end.y <span class="op">-</span><span class="st"> </span>start.y)<span class="op">/</span><span class="dv">2</span>, </span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true"></a>        <span class="dt">distance =</span> midprom <span class="op">-</span><span class="st"> </span>start.x</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(high.tss <span class="op">==</span><span class="st"> &#39;High&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true"></a><span class="st">    </span><span class="kw">count</span>(distance)</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true"></a>p&lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x =</span> distance, <span class="dt">y =</span> n)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_col</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> <span class="st">&#39;loess&#39;</span>, <span class="dt">span =</span> <span class="fl">0.05</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&#39;none&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&#39;Fragment &quot;cut&quot; sites&#39;</span>, <span class="dt">x =</span> <span class="st">&#39;Distance from TSS&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;# of cut sites&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true"></a><span class="st">    </span><span class="kw">xlim</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">1000</span>, <span class="dv">1000</span>))</span></code></pre></div>
</div>
<div id="subsetting-data" class="section level3">
<h3>Subsetting data</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a>brain<span class="op">$</span>FRiP &lt;-<span class="st"> </span>brain<span class="op">$</span>peak_region_fragments <span class="op">/</span><span class="st"> </span>brain<span class="op">$</span>passed_filters <span class="op">*</span><span class="st"> </span><span class="dv">100</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a>brain<span class="op">$</span>FRiBl &lt;-<span class="st"> </span>brain<span class="op">$</span>blacklist_region_fragments <span class="op">/</span><span class="st"> </span>brain<span class="op">$</span>peak_region_fragments</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">subset</span>(</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a>    brain,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a>    peak_region_fragments <span class="op">&gt;</span><span class="st"> </span><span class="dv">3000</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Keep cells with high number of fragments in peaks</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a><span class="st">        </span>peak_region_fragments <span class="op">&lt;</span><span class="st"> </span><span class="dv">100000</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Remove cells with number of fragments in peaks too high</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true"></a><span class="st">        </span>FRiP <span class="op">&gt;</span><span class="st"> </span><span class="dv">50</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Keep cells with high FRiP </span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true"></a><span class="st">        </span>FRiBl <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Keep cells with low FRiBl</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true"></a><span class="st">        </span>nucleosome_signal <span class="op">&lt;</span><span class="st"> </span><span class="dv">4</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Remove cells with high nucleosome signal</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true"></a><span class="st">        </span>TSS.enrichment <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span> <span class="co"># Keep cells with good TSSES</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true"></a>)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>brain[</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true"></a>    <span class="kw">rowSums</span>(<span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&quot;counts&quot;</span>)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span> <span class="op">&amp;</span><span class="st"> </span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true"></a><span class="st">    </span><span class="kw">rowSums</span>(<span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&quot;counts&quot;</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>, </span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true"></a>]</span></code></pre></div>
</div>
<div id="normalizing-and-reduce-dimensionality-data" class="section level3">
<h3>Normalizing and reduce dimensionality data</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">RunTFIDF</span>(brain) <span class="co"># Normalize data</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">FindTopFeatures</span>(brain, <span class="dt">min.cutoff =</span> <span class="st">&#39;q0&#39;</span>) <span class="co"># Find variable features</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">RunSVD</span>(<span class="dt">object =</span> brain) <span class="co"># Perform dimensionality reduction</span></span></code></pre></div>
</div>
<div id="clustering" class="section level3">
<h3>Clustering</h3>
<p>Now that cells are embedded in a lower dimensional space, they can be graph-clustered just like “regular” cells from scRNAseq.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(<span class="dt">object =</span> brain, <span class="dt">reduction =</span> <span class="st">&#39;lsi&#39;</span>, <span class="dt">dims =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">30</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> brain, <span class="dt">algorithm =</span> <span class="dv">3</span>, <span class="dt">resolution =</span> <span class="fl">1.2</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(brain, <span class="dt">reduction =</span> <span class="st">&#39;lsi&#39;</span>, <span class="dt">dims =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">30</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a>Seurat<span class="op">::</span><span class="kw">DimPlot</span>(brain, <span class="dt">group.by =</span> <span class="st">&#39;ident&#39;</span>, <span class="dt">label =</span> <span class="st">&#39;ident&#39;</span>)</span></code></pre></div>
</div>
</div>
<div id="compare-signac-processing-and-singlecellexperiment-processing" class="section level2">
<h2>2. Compare Signac processing and SingleCellExperiment processing</h2>
</div>
<div id="integrate-scatacseq-and-scrnaseq" class="section level2">
<h2>3. Integrate scATACseq and scRNAseq</h2>
<div id="download-data-1" class="section level3">
<h3>Download data</h3>
<p>scRNAseq for mouse E18 5K brain can be downloaded from
<a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K">10X Genomics</a>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a><span class="fu">mkdir</span> data/Homework_2/scRNA</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_web_summary.html -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_web_summary.html</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_metrics_summary.csv -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_metrics_summary.csv</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_cloupe.cloupe -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_cloupe.cloupe</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_alignments.bam -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_alignments.bam</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_alignments.bam.bai -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_alignments.bam.bai</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_barcodes.csv -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_barcodes.csv</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_feature_bc_matrix.h5 -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_feature_bc_matrix.h5</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_feature_bc_matrix.tar.gz -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_feature_bc_matrix.tar.gz</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_molecule_info.h5 -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_molecule_info.h5</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_analysis.tar.gz -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_analysis.tar.gz</span></code></pre></div>
</div>
</div>
<div id="functional-analysis-of-da-peaks" class="section level2">
<h2>4. Functional analysis of DA peaks</h2>
</div>
