---
title: "2. Exercises"
---

<script src="/scATACseq-workshop/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<div id="day-1" class="section level2">
<h2>Day 1</h2>
<div id="exercises-session-1" class="section level3">
<h3>Exercises session 1</h3>
<ul>
<li>Check fragment size distribution in bulk ATAC-seq</li>
</ul>
<pre class="r"><code>library(tidyverse)
library(plyranges)
fragments &lt;- VplotR::importPEBamFiles(
    &#39;?...?&#39;, 
    shift_ATAC_fragments = TRUE
)
df&lt;-?...?(fragments) %&gt;% 
    group_by(?...?) %&gt;% 
    count() %&gt;% 
    ungroup() 
p1&lt;-ggplot(df, aes(x = ?...?, y = ?...?)) + 
    ?...?() + 
    labs(title = &#39;?...?&#39;, x = &#39;?...?&#39;, y = &#39;?...?&#39;)</code></pre>
</div>
<div id="exercises-session-2" class="section level3">
<h3>Exercises session 2</h3>
<ul>
<li>Check coverage of each cluster in <code>Loupe</code> at relevant loci to propose an annotation for each cluster.</li>
</ul>
<pre class="sh"><code># Check e.g. `CD19`, `CD8`, `CD4`, `CD3`</code></pre>
</div>
</div>
<div id="day-2" class="section level2">
<h2>Day 2</h2>
<div id="exercises-session-3" class="section level3">
<h3>Exercises session 3</h3>
<ul>
<li>Inspect genes closest to the most DA peaks (in DA analysis between clusters 10+16 and 4+5). Propose a function for cells in these groups.</li>
</ul>
<pre class="r"><code>da_peaks &lt;- readRDS(&#39;data/MouseBrain/da_peaks_10-16_vs_4-5.rds&#39;)
peaks &lt;- rownames(da_peaks)[da_peaks$?...? &lt;= 0.01 &amp; da_peaks$?...? &gt;= 1.5]
peaks &lt;- GenomicRanges::GRanges(str_replace(peaks, &#39;-&#39;, &#39;:&#39;))
Annotation(brain)[nearest(?...?, Annotation(?...?))]$?...?</code></pre>
</div>
<div id="exercises-session-4" class="section level3">
<h3>Exercises session 4</h3>
<ul>
<li>Compare clustering to provided cell type activity for ce11 promoters. Comment.</li>
</ul>
<pre class="r"><code>dds &lt;- readRDS(&#39;data/ATAC_worm/quantif.rds&#39;)
rlogs &lt;- readRDS(&#39;?...?&#39;)
clusts &lt;- cluster::pam(select(rlogs, -?...?), k = 9)
table(rowData(dds)$which.tissues[!rowData(dds)$allZero], clusts$?...?)</code></pre>
</div>
</div>
<div id="day-3" class="section level2">
<h2>Day 3</h2>
<div id="exercises-session-5" class="section level3">
<h3>Exercises session 5</h3>
</div>
<div id="exercises-session-6" class="section level3">
<h3>Exercises session 6</h3>
<ul>
<li>Try different thresholds for tSNE embeddings. Comment.</li>
</ul>
<pre class="r"><code>set.seed(2021)
dev &lt;- readRDS(&#39;data/scATAC_LSC-LMPP-mono/dev.rds&#39;)
p &lt;- cowplot::plot_grid(
    chromVAR::plotDeviationsTsne(dev, chromVAR::deviationsTsne(dev, threshold = 1, perplexity = 20), annotation_name = &quot;FOS&quot;, shiny = FALSE)[[1]],
    chromVAR::plotDeviationsTsne(dev, chromVAR::deviationsTsne(dev, threshold = 1.25, perplexity = 20), annotation_name = &quot;FOS&quot;, shiny = FALSE)[[1]],
    chromVAR::plotDeviationsTsne(dev, chromVAR::deviationsTsne(dev, threshold = 1.5, perplexity = 20), annotation_name = &quot;FOS&quot;, shiny = FALSE)[[1]],
    chromVAR::plotDeviationsTsne(dev, chromVAR::deviationsTsne(dev, threshold = 1.75, perplexity = 20), annotation_name = &quot;FOS&quot;, shiny = FALSE)[[1]]
)</code></pre>
</div>
</div>
