<!DOCTYPE html>
<html lang="en" class="js csstransforms3d">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="Hugo 0.80.0" />
    <meta name="description" content="Analysis of bulk and single-cell ATAC-seq datasets">
<meta name="author" content="Jacques Serizay">

    <link rel="icon" href="/scATACseq-workshop/images/favicon.png" type="image/png">

    <title>2. Program :: Analysis of bulk and single-cell ATAC-seq datasets</title>

    
    <link href="/scATACseq-workshop/css/nucleus.css?1625303897" rel="stylesheet">
    <link href="/scATACseq-workshop/css/fontawesome-all.min.css?1625303897" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hybrid.css?1625303897" rel="stylesheet">
    <link href="/scATACseq-workshop/css/featherlight.min.css?1625303897" rel="stylesheet">
    <link href="/scATACseq-workshop/css/perfect-scrollbar.min.css?1625303897" rel="stylesheet">
    <link href="/scATACseq-workshop/css/auto-complete.css?1625303897" rel="stylesheet">
    <link href="/scATACseq-workshop/css/atom-one-dark-reasonable.css?1625303897" rel="stylesheet">
    <link href="/scATACseq-workshop/css/theme.css?1625303897" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hugo-theme.css?1625303897" rel="stylesheet">
    
    <link href="/scATACseq-workshop/css/theme-red.css?1625303897" rel="stylesheet">
    
    

    <script src="/scATACseq-workshop/js/jquery-3.3.1.min.js?1625303897"></script>

    <style>
      :root #header + #content > #left > #rlblock_left{
          display:none !important;
      }
      
    </style>
    
  </head>
  <body class="" data-url="/scATACseq-workshop/introduction/02_program/">
    <nav id="sidebar" class="">



  <div id="header-wrapper">
    <div id="header">
      <a id="logo">



</a>

    </div>
    
  </div>
  

    <div class="highlightable">
    <ul class="topics">

        
          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/introduction/" title=" Single-cell RNA-seq analysis with R/Bioconductor" class="dd-item
        parent
        
        
        ">
      <a href="/scATACseq-workshop/introduction/">
           Single-cell RNA-seq analysis with R/Bioconductor
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/01_presentation/" title="1. Presentation" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/01_presentation/">
        1. Presentation
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/02_program/" title="2. Program" class="dd-item active">
        <a href="/scATACseq-workshop/introduction/02_program/">
        2. Program
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/03_rstudio/" title="3. Rstudio" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/03_rstudio/">
        3. Rstudio
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/04_prerequisites/" title="4. Prerequisites and required machine configuration" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/04_prerequisites/">
        4. Prerequisites and required machine configuration
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/day01/" title="1. Day 01" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/day01/">
          1. Day 01
          
      </a>
      
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/extra-resources/" title="Extra resources" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/extra-resources/">
          Extra resources
          
      </a>
      
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/how_to_log_in_aws/" title="How to log in AWS" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/how_to_log_in_aws/">
          How to log in AWS
          
      </a>
      
      
    </li>
  
 

          
        
    </ul>

    
    

    
    <section id="footer">
      <center>

    <a href="https://github.com/js2264/scATACseq-workshop">Check this website source on Github</a>
    <p>Created by J. Serizay</p>
    <p>Copyright 2021</p>

</center>

<script async defer src="https://buttons.github.io/buttons.js"></script>

    </section>
  </div>
</nav>




        <section id="body">
        <div id="overlay"></div>
        <div class="padding highlightable">
        <div id="head-tags">
        
        </div>
        
        <div id="body-inner">
          
            <h1>
              
              2. Program
            </h1>
          

        


<h2 id="session-1-processing-bulk-and-single-cell-atac-seq-reads">Session 1: Processing bulk and single-cell ATAC-seq reads</h2>
<h3 id="lecture-1-bulk-atac-seq-experimental-approaches">Lecture 1: bulk ATAC-seq experimental approaches</h3>
<blockquote>
<p>15'</p>
</blockquote>
<ul>
<li>bulk ATAC-seq methods: classical, omniATAC-seq, low input ATAC-seq, &hellip;</li>
</ul>
<h3 id="hands-on-1-processing-of-bulk-atac-seq">Hands-on 1: Processing of bulk ATAC-seq</h3>
<blockquote>
<p>35'</p>
</blockquote>
<ul>
<li>Sequencing QC</li>
<li>Mapping reads and correcting</li>
<li>Visualizing and QC-ing the results</li>
</ul>
<h3 id="exercise-1-inspecting-bulk-atac-seq-outputs">Exercise 1: Inspecting bulk ATAC-seq outputs</h3>
<blockquote>
<p>10'</p>
</blockquote>
<ul>
<li>Visually inspect the generated tracks and the identifed peaks</li>
<li>Get the distribution of fragment sizes over ATAC peaks</li>
</ul>
<hr>
<h3 id="lecture-2-single-cell-atac-seq-experimental-approaches">Lecture 2: single-cell ATAC-seq experimental approaches</h3>
<blockquote>
<p>20'</p>
</blockquote>
<ul>
<li>&ldquo;single cell&rdquo; ATAC-seq methods: droplet-based scATACseq, indexed scATACseq, joint scRNAseq and scATACseq, &hellip;</li>
</ul>
<h3 id="hands-on-2-processing-of-single-cell-atac-seq">Hands-on 2: Processing of single-cell ATAC-seq</h3>
<blockquote>
<p>30'</p>
</blockquote>
<ul>
<li>10X scATACseq: from reads to matrix counts with <code>cellranger</code></li>
<li>QC-ing with <code>Loupe</code></li>
</ul>
<h3 id="exercise-2-bulk-versus-single-cell-atac-seq">Exercise 2: bulk versus single-cell ATAC-seq</h3>
<blockquote>
<p>10'</p>
</blockquote>
<ul>
<li>Compare aggregated signal obtained from single-cell ATAC-seq to that of bulk ATAC-seq</li>
</ul>
<hr>
<h3 id="homework-download-raw-data-from-gse-and-process-it">Homework: Download raw data from GSE and process it</h3>
<ul>
<li>Install <code>cellranger</code> and mapping tools</li>
<li>Download reads from GSE129785 and GSE87646</li>
</ul>
<h2 id="session-2-peak-centered-differential-accessibility-analysis">Session 2: Peak-centered differential accessibility analysis</h2>
<h3 id="lecture-3-clustering-single-cell-data">Lecture 3: Clustering single-cell data</h3>
<blockquote>
<p>20'</p>
</blockquote>
<ul>
<li>Normalizing scATACseq data</li>
<li>Clustering approaches</li>
<li>Annotating cell types</li>
</ul>
<h3 id="hands-on-3-downstream-analysis-of-scatacseq-data">Hands-on 3: Downstream analysis of scATACseq data</h3>
<blockquote>
<p>30'</p>
</blockquote>
<ul>
<li>Reading <code>cellranger</code> output in <code>R</code></li>
<li>Normalizing scATACseq counts</li>
<li>Clustering cells by their accessible loci</li>
</ul>
<h3 id="exercise-3-recovering-peaks-associated-with">Exercise 3: Recovering peaks associated with</h3>
<blockquote>
<p>10'</p>
</blockquote>
<ul>
<li>Finding cluster-specific peaks</li>
</ul>
<hr>
<h3 id="lecture-4-atac-seq-and-accessibility-peaks">Lecture 4: ATAC-seq and accessibility peaks</h3>
<blockquote>
<p>20'</p>
</blockquote>
<ul>
<li>Peaks in bulk ATAC-seq vs single-cell ATAC-seq</li>
<li>Integrating peak sets from different experiments</li>
<li>Finding enriched sequences</li>
</ul>
<h3 id="hands-on-4-comparing-atac-seq-peaks">Hands-on 4: Comparing ATAC-seq peaks</h3>
<blockquote>
<p>30'</p>
</blockquote>
<ul>
<li>Quantifying accessibility at ATAC peaks</li>
<li>Differential accessibility assay</li>
<li>Clustering peak sets</li>
</ul>
<h3 id="exercise-4-peak-location-vs-genomic-features">Exercise 4: Peak location vs genomic features</h3>
<blockquote>
<p>10'</p>
</blockquote>
<ul>
<li>Get the distance from ATAC peaks to closest TSS</li>
<li>Find which types of chromatin stats are enriched in acceessible loci</li>
</ul>
<hr>
<h3 id="homework">Homework</h3>
<ul>
<li>Group peaks by accessibility scores in different samples / cells</li>
<li>Perform gene ontology over-representation analyses (GOA)</li>
<li>Perform gene Set Enrichment Analyses (GSEA)</li>
</ul>
<h2 id="session-3-advanced-topics-and-multi-omics-integration">Session 3: Advanced topics and multi-omics integration</h2>
<h3 id="lecture-5-motif-enrichment-analysis">Lecture 5: Motif enrichment analysis</h3>
<blockquote>
<p>20'</p>
</blockquote>
<ul>
<li>The good, the bad and the ugly: Ref. sequence, ref. annotations and ref. version!</li>
<li>Tools to perform motif enrichment analysis</li>
<li>Public databases for transcription factor binding sites (TFBS)</li>
</ul>
<h3 id="hands-on-5-find-motifs-enriched-in-peak-sets">Hands-on 5: Find motifs enriched in peak sets</h3>
<blockquote>
<p>30'</p>
</blockquote>
<ul>
<li>Use web-based or command-line <code>meme</code> software to find de novo motifs enriched in a set of peaks</li>
<li>Scan sets of peaks for known TF binding sites</li>
</ul>
<h3 id="hands-on-5-find-motifs-enriched-in-peak-sets-1">Hands-on 5: Find motifs enriched in peak sets</h3>
<blockquote>
<p>10'</p>
</blockquote>
<ul>
<li>Compare de novo found motifs to public databases with <code>fimo</code></li>
</ul>
<hr>
<h3 id="topic-6-inferring-regulatory-networks-from-scatacseq">Topic 6: Inferring regulatory networks from scATACseq</h3>
<blockquote>
<p>1h'</p>
</blockquote>
<ul>
<li>CICERO</li>
</ul>
<hr>
<h3 id="homework-1">Homework</h3>
<ul>
<li>Find TF motifs enriched in peaks with increased accessibility in a specific cluster</li>
<li>Find co-regulated genes using CICERO</li>
</ul>


<footer class="footline">
	
</footer>

        
        </div>
        

      </div>

    <div id="navigation">
        
        

        
            
            
                
                    
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        

        


	 
	 
    </div>

    </section>

    <div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
      <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
    </div>
    <script src="/scATACseq-workshop/js/clipboard.min.js?1625303897"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.min.js?1625303897"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.jquery.min.js?1625303897"></script>
    <script src="/scATACseq-workshop/js/jquery.sticky.js?1625303897"></script>
    <script src="/scATACseq-workshop/js/featherlight.min.js?1625303897"></script>
    <script src="/scATACseq-workshop/js/highlight.pack.js?1625303897"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="/scATACseq-workshop/js/modernizr.custom-3.6.0.js?1625303897"></script>
    <script src="/scATACseq-workshop/js/learn.js?1625303897"></script>
    <script src="/scATACseq-workshop/js/hugo-learn.js?1625303897"></script>
    
        
            <script src="/scATACseq-workshop/mermaid/mermaid.js?1625303897"></script>
        
        <script>
            mermaid.initialize({ startOnLoad: true });
        </script>
    
    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-105947713-1', 'auto');
  ga('send', 'pageview');

</script>

  </body>
</html>

