<script src="_index_files/header-attrs-2.8/header-attrs.js"></script>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<div id="day-1" class="section level2">
<h2>Day 1</h2>
<div id="exercises-session-1" class="section level3">
<h3>Exercises session 1</h3>
<ul>
<li>Check fragment size distribution in bulk ATAC-seq</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="kw">library</span>(plyranges)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a>fragments &lt;-<span class="st"> </span>VplotR<span class="op">::</span><span class="kw">importPEBamFiles</span>(</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a>    <span class="st">&#39;?...?&#39;</span>, </span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a>    <span class="dt">shift_ATAC_fragments =</span> <span class="ot">TRUE</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a>df&lt;-?...?(fragments) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">group_by</span>(?...?) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">count</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true"></a><span class="st">    </span><span class="kw">ungroup</span>() </span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true"></a>p1&lt;-<span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x =</span> ?...?, <span class="dt">y =</span> ?...?)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true"></a><span class="st">    </span>?...?() <span class="op">+</span><span class="st"> </span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&#39;?...?&#39;</span>, <span class="dt">x =</span> <span class="st">&#39;?...?&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;?...?&#39;</span>)</span></code></pre></div>
</div>
<div id="exercises-session-2" class="section level3">
<h3>Exercises session 2</h3>
<ul>
<li>Check coverage of each cluster in <code>Loupe</code> at relevant loci to propose an annotation for each cluster.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="co"># Check e.g. `CD19`, `CD8`, `CD4`, `CD3`</span></span></code></pre></div>
</div>
</div>
<div id="day-2" class="section level2">
<h2>Day 2</h2>
<div id="exercises-session-3" class="section level3">
<h3>Exercises session 3</h3>
<ul>
<li>Inspect genes closest to the most DA peaks (in DA analysis between clusters 10+16 and 4+5). Propose a function for cells in these groups.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a>da_peaks &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&#39;data/MouseBrain/da_peaks_10-16_vs_4-5.rds&#39;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a>peaks &lt;-<span class="st"> </span><span class="kw">rownames</span>(da_peaks)[da_peaks<span class="op">$</span>?...? <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.01</span> <span class="op">&amp;</span><span class="st"> </span>da_peaks<span class="op">$</span>?...? <span class="op">&gt;=</span><span class="st"> </span><span class="fl">1.5</span>]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a>peaks &lt;-<span class="st"> </span>GenomicRanges<span class="op">::</span><span class="kw">GRanges</span>(<span class="kw">str_replace</span>(peaks, <span class="st">&#39;-&#39;</span>, <span class="st">&#39;:&#39;</span>))</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a><span class="kw">Annotation</span>(brain)[<span class="kw">nearest</span>(?...?, <span class="kw">Annotation</span>(?...?))]<span class="op">$</span>?...?</span></code></pre></div>
</div>
<div id="exercises-session-4" class="section level3">
<h3>Exercises session 4</h3>
<ul>
<li>Compare clustering to provided cell type activity for ce11 promoters. Comment.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a>dds &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&#39;data/ATAC_worm/quantif.rds&#39;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a>rlogs &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&#39;?...?&#39;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a>clusts &lt;-<span class="st"> </span>cluster<span class="op">::</span><span class="kw">pam</span>(<span class="kw">select</span>(rlogs, <span class="op">-</span>?...?), <span class="dt">k =</span> <span class="dv">9</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a><span class="kw">table</span>(<span class="kw">rowData</span>(dds)<span class="op">$</span>which.tissues[<span class="op">!</span><span class="kw">rowData</span>(dds)<span class="op">$</span>allZero], clusts<span class="op">$</span>?...?)</span></code></pre></div>
</div>
</div>
<div id="day-3" class="section level2">
<h2>Day 3</h2>
<div id="exercises-session-5" class="section level3">
<h3>Exercises session 5</h3>
</div>
<div id="exercises-session-6" class="section level3">
<h3>Exercises session 6</h3>
<ul>
<li>Try different thresholds for tSNE embeddings. Comment.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="kw">set.seed</span>(<span class="dv">2021</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a>dev &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&#39;data/scATAC_LSC-LMPP-mono/dev.rds&#39;</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a>p &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw">plot_grid</span>(</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a>    chromVAR<span class="op">::</span><span class="kw">plotDeviationsTsne</span>(dev, chromVAR<span class="op">::</span><span class="kw">deviationsTsne</span>(dev, <span class="dt">threshold =</span> <span class="dv">1</span>, <span class="dt">perplexity =</span> <span class="dv">20</span>), <span class="dt">annotation_name =</span> <span class="st">&quot;FOS&quot;</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)[[<span class="dv">1</span>]],</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a>    chromVAR<span class="op">::</span><span class="kw">plotDeviationsTsne</span>(dev, chromVAR<span class="op">::</span><span class="kw">deviationsTsne</span>(dev, <span class="dt">threshold =</span> <span class="fl">1.25</span>, <span class="dt">perplexity =</span> <span class="dv">20</span>), <span class="dt">annotation_name =</span> <span class="st">&quot;FOS&quot;</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)[[<span class="dv">1</span>]],</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true"></a>    chromVAR<span class="op">::</span><span class="kw">plotDeviationsTsne</span>(dev, chromVAR<span class="op">::</span><span class="kw">deviationsTsne</span>(dev, <span class="dt">threshold =</span> <span class="fl">1.5</span>, <span class="dt">perplexity =</span> <span class="dv">20</span>), <span class="dt">annotation_name =</span> <span class="st">&quot;FOS&quot;</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)[[<span class="dv">1</span>]],</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true"></a>    chromVAR<span class="op">::</span><span class="kw">plotDeviationsTsne</span>(dev, chromVAR<span class="op">::</span><span class="kw">deviationsTsne</span>(dev, <span class="dt">threshold =</span> <span class="fl">1.75</span>, <span class="dt">perplexity =</span> <span class="dv">20</span>), <span class="dt">annotation_name =</span> <span class="st">&quot;FOS&quot;</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)[[<span class="dv">1</span>]]</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true"></a>)</span></code></pre></div>
</div>
</div>
