<!DOCTYPE html>
<html lang="en" class="js csstransforms3d">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="Hugo 0.80.0" />
    <meta name="description" content="Analysis of bulk and single-cell ATAC-seq datasets">
<meta name="author" content="Jacques Serizay">

    <link rel="icon" href="/scATACseq-workshop/images/favicon.png" type="image/png">

    <title>Day 01 :: Analysis of bulk and single-cell ATAC-seq datasets</title>

    
    <link href="/scATACseq-workshop/css/nucleus.css?1628535635" rel="stylesheet">
    <link href="/scATACseq-workshop/css/fontawesome-all.min.css?1628535635" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hybrid.css?1628535635" rel="stylesheet">
    <link href="/scATACseq-workshop/css/featherlight.min.css?1628535635" rel="stylesheet">
    <link href="/scATACseq-workshop/css/perfect-scrollbar.min.css?1628535635" rel="stylesheet">
    <link href="/scATACseq-workshop/css/auto-complete.css?1628535635" rel="stylesheet">
    <link href="/scATACseq-workshop/css/atom-one-dark-reasonable.css?1628535635" rel="stylesheet">
    <link href="/scATACseq-workshop/css/theme.css?1628535635" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hugo-theme.css?1628535635" rel="stylesheet">
    
    <link href="/scATACseq-workshop/css/theme-red.css?1628535635" rel="stylesheet">
    
    <link href="/scATACseq-workshop/css/custom_css.css?1628535635" rel="stylesheet">

    <script src="/scATACseq-workshop/js/jquery-3.3.1.min.js?1628535635"></script>

    <style>
      :root #header + #content > #left > #rlblock_left{
          display:none !important;
      }
      
        :not(pre) > code + span.copy-to-clipboard {
            display: none;
        }
      
    </style>
    
  </head>
  <body class="" data-url="/scATACseq-workshop/hands-on-sessions/day1/">
    <nav id="sidebar" class="">



  <div id="header-wrapper">
    <div id="header">
      <a id="logo">



</a>

    </div>
    
  </div>
  

    <div class="highlightable">
    <ul class="topics">

        
          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/introduction/" title=" Bulk and single-cell ATAC-seq analysis" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/introduction/">
           Bulk and single-cell ATAC-seq analysis
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/02_program/" title="1. Program" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/02_program/">
        1. Program
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/01_prerequisites/" title="2. Prerequisites and required machine configuration" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/01_prerequisites/">
        2. Prerequisites and required machine configuration
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/hands-on-sessions/" title="1. Hands-on sessions" class="dd-item
        parent
        
        
        ">
      <a href="/scATACseq-workshop/hands-on-sessions/">
          1. Hands-on sessions
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands-on-sessions/day1/" title="Day 01" class="dd-item active">
        <a href="/scATACseq-workshop/hands-on-sessions/day1/">
        Day 01
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands-on-sessions/day2/" title="Day 02" class="dd-item ">
        <a href="/scATACseq-workshop/hands-on-sessions/day2/">
        Day 02
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands-on-sessions/day3/" title="Day 03" class="dd-item ">
        <a href="/scATACseq-workshop/hands-on-sessions/day3/">
        Day 03
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/exercises/" title="2. Exercises" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/exercises/">
          2. Exercises
          
      </a>
      
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/homeworks/" title="3. Homeworks" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/homeworks/">
          3. Homeworks
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_1/" title="Homework 1" class="dd-item ">
        <a href="/scATACseq-workshop/homeworks/homework_1/">
        Homework 1
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_2/" title="Homework 2" class="dd-item ">
        <a href="/scATACseq-workshop/homeworks/homework_2/">
        Homework 2
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_3/" title="Homework 3" class="dd-item ">
        <a href="/scATACseq-workshop/homeworks/homework_3/">
        Homework 3
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/extra-resources/" title="Extra resources" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/extra-resources/">
          Extra resources
          
      </a>
      
      
    </li>
  
 

          
        
    </ul>

    
    

    
    <section id="footer">
      <center>

    <a href="https://github.com/js2264/scATACseq-workshop">Check this website source on Github</a>
    <p>Created by J. Serizay</p>
    <p>Copyright 2021</p>

</center>

<script async defer src="https://buttons.github.io/buttons.js"></script>

    </section>
  </div>
</nav>




        <section id="body">
        <div id="overlay"></div>
        <div class="padding highlightable">
        <div id="head-tags">
        
        </div>
        
        <div id="body-inner">
          
            <h1>
              
              Day 01
            </h1>
          

        


<h2 id="hands-on-session-1-processing-bulk-atac-seq">Hands-on session 1: Processing bulk ATAC-seq</h2>
<h3 id="download-fastq">Download fastq</h3>
<p>Here we will process one of the first samples generated by ATAC-seq, from (doi: 10.1038/nmeth.2688)[https://doi.org/10.1038/nmeth.2688].
This is an ATAC-seq sample from GM12878 cells. Let&rsquo;s first download the raw fastqs.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">mkdir -p data/GM12878/fastq
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR891/SRR891269/SRR891269_1.fastq.gz -O data/GM12878/fastq/GM12878_R1.fq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR891/SRR891269/SRR891269_2.fastq.gz -O data/GM12878/fastq/GM12878_R2.fq.gz
</code></pre></div><h3 id="process-gm12878-data">Process GM12878 data</h3>
<p>The processing steps for ATAC-seq are:</p>
<ol>
<li>Adapter trimming</li>
<li>Read pairs mapping</li>
<li>Read pairs filtering</li>
<li>Track generation</li>
<li>Peak calling</li>
</ol>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">cd data/GM12878
SAMPLE<span style="color:#f92672">=</span>GM12878
R1<span style="color:#f92672">=</span>fastq/GM12878_R1.fq.gz
R2<span style="color:#f92672">=</span>fastq/GM12878_R2.fq.gz
GENOME<span style="color:#f92672">=</span>~/genomes/hg38/hg38.fa
SAMFILE<span style="color:#f92672">=</span>GM12878.sam
FILTEREDBAM<span style="color:#f92672">=</span>GM12878_filtered.bam
TRACK<span style="color:#f92672">=</span>GM12878.bw
PEAKFOLDER<span style="color:#f92672">=</span>peaks/GM12878
GENOMESIZE<span style="color:#f92672">=</span><span style="color:#ae81ff">2600000000</span>
CPU<span style="color:#f92672">=</span><span style="color:#ae81ff">2</span>

<span style="color:#75715e">## -- Trim reads and run QC</span>
trim_galore <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --paired <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --fastqc <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --cores <span style="color:#ae81ff">1</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>R1<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>R2<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span>

<span style="color:#75715e">## -- Map reads </span>
bowtie2 <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --threads <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    -x <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>GENOME<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --maxins <span style="color:#ae81ff">2000</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    -1 <span style="color:#e6db74">`</span>echo <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>R1<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> | sed <span style="color:#e6db74">&#39;s,.fastq.gz,_val_1.fq.gz,&#39;</span><span style="color:#e6db74">`</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    -2 <span style="color:#e6db74">`</span>echo <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>R2<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> | sed <span style="color:#e6db74">&#39;s,.fastq.gz,_val_2.fq.gz,&#39;</span><span style="color:#e6db74">`</span> &gt; <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>SAMFILE<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span>

<span style="color:#75715e">## -- Filter reads </span>
samtools fixmate -@ <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> --output-fmt bam -m <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>SAMFILE<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> - | <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>samtools sort -@ <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> --output-fmt bam -T <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>SAMFILE<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span>_sorting - | <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>samtools markdup -@ <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> --output-fmt bam -r -T <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>SAMFILE<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span>_markdup - - | <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>samtools view -@ <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> --output-fmt bam -f <span style="color:#ae81ff">2</span> -q <span style="color:#ae81ff">10</span> -1 -b - | <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>samtools sort -@ <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> --output-fmt bam -l <span style="color:#ae81ff">9</span> -T <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>SAMFILE<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span>_sorting2 -o <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>FILTEREDBAM<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span>
samtools index -@ <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>FILTEREDBAM<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span>

<span style="color:#75715e">## -- Generate track</span>
bamCoverage <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --bam <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>FILTEREDBAM<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --outFileName <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>TRACK<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --binSize <span style="color:#ae81ff">1</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --numberOfProcessors <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --normalizeUsing CPM <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --skipNonCoveredRegions <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --extendReads <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --ignoreDuplicates

<span style="color:#75715e">## -- Call peaks</span>
macs2 callpeak <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    -t <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>FILTEREDBAM<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --format BAMPE <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --gsize <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>GENOMESIZE<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --outdir <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>PEAKFOLDER<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --name <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>SAMPLE<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span>
</code></pre></div><h3 id="qc-of-gm12878-data">QC of GM12878 data</h3>
<p>To check the quality of an ATAC-seq sample, one can check the following metrics:</p>
<ol>
<li>Fragment size distribution</li>
<li>Distance from fragment to TSS</li>
<li>1+2 combined: Vplots</li>
<li>TSS enrichment score (coverage enrichment at TSSs)</li>
<li>Location of peaks vs. genomic features</li>
<li>Fraction of fragments within peaks</li>
</ol>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-r" data-lang="r"><span style="color:#a6e22e">library</span>(AnnotationDbi)
<span style="color:#a6e22e">library</span>(tidyverse)
<span style="color:#a6e22e">library</span>(plyranges)
<span style="color:#a6e22e">library</span>(annotatr)
<span style="color:#a6e22e">library</span>(VplotR)

<span style="color:#75715e">## -- Import genomic features</span>
hg_genes <span style="color:#f92672">&lt;-</span> AnnotationHub<span style="color:#f92672">::</span><span style="color:#a6e22e">AnnotationHub</span>()[[<span style="color:#e6db74">&#39;AH92109&#39;</span>]] <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">filter</span>(type <span style="color:#f92672">==</span> <span style="color:#e6db74">&#39;gene&#39;</span>, gene_biotype <span style="color:#f92672">==</span> <span style="color:#e6db74">&#39;protein_coding&#39;</span>) <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">filter</span>(<span style="color:#f92672">!</span>(<span style="color:#a6e22e">grepl</span>(<span style="color:#e6db74">&#39;GL|KI&#39;</span>, seqnames)))
<span style="color:#a6e22e">seqlevelsStyle</span>(hg_genes) <span style="color:#f92672">&lt;-</span> <span style="color:#e6db74">&#39;UCSC&#39;</span>
hg_TSSs <span style="color:#f92672">&lt;-</span> hg_genes <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">anchor_start</span>() <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">resize</span>(width <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>) <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">unanchor</span>()
hg_promoters <span style="color:#f92672">&lt;-</span> hg_TSSs <span style="color:#f92672">%&gt;%</span>
    <span style="color:#a6e22e">flank_upstream</span>(<span style="color:#ae81ff">4000</span>) <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">shift_downstream</span>(<span style="color:#ae81ff">2000</span>) <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">mutate</span>(ID <span style="color:#f92672">=</span> <span style="color:#a6e22e">paste0</span>(<span style="color:#e6db74">&#39;prom_&#39;</span>, <span style="color:#ae81ff">1</span><span style="color:#f92672">:</span><span style="color:#a6e22e">n</span>()))
hg_features <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">build_annotations</span>(
    genome <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;hg38&#39;</span>,
    annotations <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;hg38_basicgenes&#39;</span>
)

<span style="color:#75715e">## -- Import fragments and peaks as GRanges</span>
fragments <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">importPEBamFiles</span>(
    <span style="color:#e6db74">&#39;GM12878_filtered.bam&#39;</span>, 
    shift_ATAC_fragments <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>
)
peaks <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">import</span>(<span style="color:#e6db74">&#39;peaks/GM12878/GM12878_peaks.narrowPeak&#39;</span>)
peak_summits <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">import</span>(<span style="color:#e6db74">&#39;peaks/GM12878/GM12878_summits.bed&#39;</span>)

<span style="color:#75715e">## -- Check fragment size distribution</span>
df<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">...</span>(fragments) <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">group_by</span>(<span style="color:#66d9ef">...</span>) <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">count</span>() <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">ungroup</span>() 
p1<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">ggplot</span>(df, <span style="color:#a6e22e">aes</span>(x <span style="color:#f92672">=</span> <span style="color:#66d9ef">...</span>, y <span style="color:#f92672">=</span> <span style="color:#66d9ef">...</span>)) <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">...</span>() <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">labs</span>(title <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;...&#39;</span>, x <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;...&#39;</span>, y <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;...&#39;</span>)

<span style="color:#75715e">## -- Check distance to TSS</span>
df<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">add_nearest_distance</span>(fragments, hg_TSSs) <span style="color:#f92672">%&gt;%</span>
    <span style="color:#a6e22e">as_tibble</span>() <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">mutate</span>(distance <span style="color:#f92672">=</span> <span style="color:#a6e22e">cut</span>(distance, breaks <span style="color:#f92672">=</span> <span style="color:#a6e22e">seq</span>(<span style="color:#ae81ff">0</span>, <span style="color:#ae81ff">5e5</span>, by <span style="color:#f92672">=</span> <span style="color:#ae81ff">10</span>), include.lowest <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>) <span style="color:#f92672">%&gt;%</span> <span style="color:#a6e22e">as.numeric</span>() <span style="color:#f92672">%&gt;%</span> <span style="color:#a6e22e">`*`</span>(<span style="color:#ae81ff">10</span>)) <span style="color:#f92672">%&gt;%</span>
    <span style="color:#a6e22e">group_by</span>(distance) <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">count</span>() <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">ungroup</span>() <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">mutate</span>(cumsum <span style="color:#f92672">=</span> <span style="color:#a6e22e">cumsum</span>(n), pct <span style="color:#f92672">=</span> cumsum<span style="color:#f92672">/</span><span style="color:#a6e22e">max</span>(cumsum))
p2<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">ggplot</span>(df, <span style="color:#a6e22e">aes</span>(x <span style="color:#f92672">=</span> distance, y <span style="color:#f92672">=</span> pct)) <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">geom_line</span>() <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">theme_minimal</span>() <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">scale_x_log10</span>() <span style="color:#f92672">+</span>
    <span style="color:#a6e22e">theme</span>(legend.position <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;none&#39;</span>) <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">labs</span>(title <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Cumulative distribution of ATAC-seq fragments&#39;</span>, x <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Distance from TSS&#39;</span>, y <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Cum. %&#39;</span>)

<span style="color:#75715e">## -- Check Vplot</span>
p3<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">plotVmat</span>(
    fragments, 
    hg_TSSs, 
    normFun <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;zscore&#39;</span>, 
    ylim <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#ae81ff">20</span>, <span style="color:#ae81ff">800</span>), 
    xlim <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#ae81ff">-500</span>, <span style="color:#ae81ff">500</span>)
)

<span style="color:#75715e">## -- Compute promoter (TSS) Enrichment Score</span>
df<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">resize</span>(fragments, width <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>, fix <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;center&#39;</span>) <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">join_overlap_left</span>(hg_promoters) <span style="color:#f92672">%&gt;%</span> 
    plyranges<span style="color:#f92672">::</span><span style="color:#a6e22e">select</span>(ID) <span style="color:#f92672">%&gt;%</span>
    <span style="color:#a6e22e">as_tibble</span>() <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">filter</span>(<span style="color:#f92672">!</span><span style="color:#a6e22e">is.na</span>(ID)) <span style="color:#f92672">%&gt;%</span>
    <span style="color:#a6e22e">left_join</span>(<span style="color:#a6e22e">as_tibble</span>(hg_promoters) <span style="color:#f92672">%&gt;%</span> dplyr<span style="color:#f92672">::</span><span style="color:#a6e22e">select</span>(start, end, ID), by <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;ID&#39;</span>) <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">mutate</span>(
        prom_mid <span style="color:#f92672">=</span> start.y <span style="color:#f92672">+</span> (end.y <span style="color:#f92672">-</span> start.y) <span style="color:#f92672">/</span> <span style="color:#ae81ff">2</span>, 
        distance <span style="color:#f92672">=</span> start.x <span style="color:#f92672">-</span> prom_mid, 
        binned_distance <span style="color:#f92672">=</span> <span style="color:#a6e22e">round</span>(distance<span style="color:#f92672">/</span><span style="color:#ae81ff">100</span>, <span style="color:#ae81ff">0</span>)<span style="color:#f92672">*</span><span style="color:#ae81ff">100</span>
    ) <span style="color:#f92672">%&gt;%</span>
    <span style="color:#a6e22e">count</span>(binned_distance) <span style="color:#f92672">%&gt;%</span>
    <span style="color:#a6e22e">mutate</span>(
        bg <span style="color:#f92672">=</span> <span style="color:#a6e22e">mean</span>(.data<span style="color:#f92672">$</span>n<span style="color:#a6e22e">[abs</span>(.data<span style="color:#f92672">$</span>binned_distance) <span style="color:#f92672">==</span> <span style="color:#ae81ff">2000</span>]), 
        enrich_score <span style="color:#f92672">=</span> n<span style="color:#f92672">/</span>bg
    )
TSSES <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">round</span>(df[df<span style="color:#f92672">$</span>binned_distance <span style="color:#f92672">==</span> <span style="color:#ae81ff">0</span>, <span style="color:#e6db74">&#39;enrich_score&#39;</span>], <span style="color:#ae81ff">2</span>)
p4<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">ggplot</span>(df, <span style="color:#a6e22e">aes</span>(x <span style="color:#f92672">=</span> binned_distance, y <span style="color:#f92672">=</span> enrich_score)) <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">geom_line</span>() <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">theme_minimal</span>() <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">theme</span>(legend.position <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;none&#39;</span>) <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">labs</span>(title <span style="color:#f92672">=</span> glue<span style="color:#f92672">::</span><span style="color:#a6e22e">glue</span>(<span style="color:#e6db74">&#39;TSS enrichment score (Signal-to-noise ratio): {TSSES}&#39;</span>), x <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Distance from TSS&#39;</span>, y <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Enrich. score&#39;</span>)

<span style="color:#75715e">## -- Check location of peaks vs genomic features</span>
peaks_annotated <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">annotate_regions</span>(
    peak_summits, 
    annotations <span style="color:#f92672">=</span> hg_features,
    ignore.strand <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>
)
p5<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">plot_annotation</span>(
    annotated_regions <span style="color:#f92672">=</span> peaks_annotated,
    annotation_order <span style="color:#f92672">=</span> <span style="color:#a6e22e">paste0</span>(<span style="color:#e6db74">&#39;hg38_genes_&#39;</span>, <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#39;promoters&#39;</span>, <span style="color:#e6db74">&#39;1to5kb&#39;</span>, <span style="color:#e6db74">&#39;5UTRs&#39;</span>, <span style="color:#e6db74">&#39;exons&#39;</span>, <span style="color:#e6db74">&#39;introns&#39;</span>, <span style="color:#e6db74">&#39;3UTRs&#39;</span>)),
    plot_title <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;ATAC peaks&#39;</span>,
    x_label <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Genomic features&#39;</span>,
    y_label <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Count&#39;</span>
) <span style="color:#f92672">+</span> <span style="color:#a6e22e">theme_minimal</span>()

<span style="color:#75715e">## -- Compute FRiP </span>
df<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">add_nearest_distance</span>(fragments, peaks) <span style="color:#f92672">%&gt;%</span>
    <span style="color:#a6e22e">as_tibble</span>() <span style="color:#f92672">%&gt;%</span> 
    <span style="color:#a6e22e">mutate</span>(isInPeak <span style="color:#f92672">=</span> distance <span style="color:#f92672">==</span> <span style="color:#ae81ff">0</span>)
pct <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">round</span>(<span style="color:#a6e22e">sum</span>(df<span style="color:#f92672">$</span>isInPeak <span style="color:#f92672">==</span> <span style="color:#ae81ff">1</span>) <span style="color:#f92672">/</span> <span style="color:#a6e22e">nrow</span>(df) <span style="color:#f92672">*</span> <span style="color:#ae81ff">100</span>, <span style="color:#ae81ff">2</span>)
p6<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">ggplot</span>(df, <span style="color:#a6e22e">aes</span>(y <span style="color:#f92672">=</span> isInPeak)) <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">geom_bar</span>() <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">theme_minimal</span>() <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">theme</span>(legend.position <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;none&#39;</span>) <span style="color:#f92672">+</span> 
    <span style="color:#a6e22e">labs</span>(title <span style="color:#f92672">=</span> glue<span style="color:#f92672">::</span><span style="color:#a6e22e">glue</span>(<span style="color:#e6db74">&#39;Fraction of reads in peaks (FRiP): {pct}&#39;</span>), x <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;# of fragments&#39;</span>, y <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;Fragment within peak&#39;</span>)

<span style="color:#75715e">## -- All plots together</span>
p <span style="color:#f92672">&lt;-</span> cowplot<span style="color:#f92672">::</span><span style="color:#a6e22e">plot_grid</span>(
    p1, p2, p3, p4, p5, p6, 
    nrow <span style="color:#f92672">=</span> <span style="color:#ae81ff">2</span>
)
<span style="color:#a6e22e">ggsave</span>(<span style="color:#e6db74">&#39;ATACseq_QC.pdf&#39;</span>, w <span style="color:#f92672">=</span> <span style="color:#ae81ff">15</span>, h <span style="color:#f92672">=</span> <span style="color:#ae81ff">10</span>)
</code></pre></div><h2 id="hands-on-session-2-processing-single-cell-atac-seq">Hands-on session 2: Processing single-cell ATAC-seq</h2>
<p>Single-cell ATAC-seq data processing can be performed just like bulk ATAC-seq, as long as the
sequencing strategy is similar (e.g. in ATAC-seq from individually sorted cells).
However, 10X genomics does not follow the regular sequencing strategy used for standard
Illumina libraries, as it needs to sequenced internal barcodes. 10X genomics has come up
with a toolkit to process 10X scATACseq data into a single <code>bam</code> file, to identify peaks of
chromatin accessibility in the entire set of single cells, and to count fragments overlapping
with each peak in each cell.</p>
<h3 id="download-fastq-of-10x-provided--1000-human-pbmcs">Download fastq of 10X-provided ~ 1,000 human PBMCs</h3>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">mkdir -p data/PBMCs_10X/
wget https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_pbmc_1k_v1/atac_pbmc_1k_v1_fastqs.tar -O data/PBMCs_10X/atac_pbmc_1k_v1_fastqs.tar.gz
tar -xf data/PBMCs_10X/atac_pbmc_1k_v1_fastqs.tar.gz
mv atac_pbmc_1k_v1_fastqs data/PBMCs_10X/fastqs
</code></pre></div><h3 id="perform-cellranger-atac-count-job-on-a-hpc-slurm">Perform cellranger-atac count job on a HPC (Slurm)</h3>
<p>It is (most likely) necessary to run <code>cellranger-atac count</code> workflow from an HPC node. Requirements are
very high (see here for more details: <a href="https://support.10xgenomics.com/single-cell-atac/software/overview/system-requirements">https://support.10xgenomics.com/single-cell-atac/software/overview/system-requirements</a>)
and processing times can be quite long. It is a completely different workflow than the one used for 10X scRNAseq!!</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">CPU<span style="color:#f92672">=</span><span style="color:#ae81ff">48</span>
MEM<span style="color:#f92672">=</span>256G
salloc <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    -J cellranger_atac_count <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --cpus-per-task <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --mem <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>MEM<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> 
</code></pre></div><p>Once the allocated node is ready:</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">CPU<span style="color:#f92672">=</span><span style="color:#ae81ff">48</span>
MEM<span style="color:#f92672">=</span>256G
ID<span style="color:#f92672">=</span>PBMCs_10X
REF<span style="color:#f92672">=</span>~/genomes/refdata-cellranger-arc-hg19-2020-A-2.0.0/
FASTQS<span style="color:#f92672">=</span>data/PBMCs_10X/fastqs
module load cellranger-atac
cellranger-atac count <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --id <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>ID<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --reference <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>REF<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --fastqs <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>FASTQS<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --localcores <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --localmem <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>MEM<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span>
</code></pre></div><p>Or, to submit the job itself from main server:</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">CPU<span style="color:#f92672">=</span><span style="color:#ae81ff">48</span>
MEM<span style="color:#f92672">=</span>256G
ID<span style="color:#f92672">=</span>PBMCs_10X
REF<span style="color:#f92672">=</span>~/genomes/refdata-cellranger-arc-hg19-2020-A-2.0.0/
FASTQS<span style="color:#f92672">=</span>data/PBMCs_10X/fastqs
sbatch <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    -J cellranger_atac_count <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    -o cellranger_atac_count.pbmc.out <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    -e cellranger_atac_count.pbmc.err <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --cpus-per-task <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>CPU<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --mem <span style="color:#e6db74">&#34;</span><span style="color:#e6db74">${</span>MEM<span style="color:#e6db74">}</span><span style="color:#e6db74">&#34;</span> <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --export<span style="color:#f92672">=</span>ALL <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    bin/submit_cellranger-atac-count.sh
</code></pre></div><p>Once the analysis is done, you can check the generated files and QC reports.
See here for more details: <a href="https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/using/count/">https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/using/count/</a>.
Our QC file is in:</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-sh" data-lang="sh">data/PBMCs_10X/atac_pbmc_1k_v1_web_summary.html
</code></pre></div>

<footer class="footline">
	
</footer>

        
        </div>
        

      </div>

    <div id="navigation">
        
        

        
            
            
                
                    
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        

        


	 
	 
		
			<a class="nav nav-prev" href="/scATACseq-workshop/hands-on-sessions/" title="1. Hands-on sessions"> <i class="fa fa-chevron-left"></i></a>
		
		
			<a class="nav nav-next" href="/scATACseq-workshop/hands-on-sessions/day2/" title="Day 02" style="margin-right: 0px;"><i class="fa fa-chevron-right"></i></a>
		
	
    </div>

    </section>

    <div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
      <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
    </div>
    <script src="/scATACseq-workshop/js/clipboard.min.js?1628535635"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.min.js?1628535635"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.jquery.min.js?1628535635"></script>
    <script src="/scATACseq-workshop/js/jquery.sticky.js?1628535635"></script>
    <script src="/scATACseq-workshop/js/featherlight.min.js?1628535635"></script>
    <script src="/scATACseq-workshop/js/highlight.pack.js?1628535635"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="/scATACseq-workshop/js/modernizr.custom-3.6.0.js?1628535635"></script>
    <script src="/scATACseq-workshop/js/learn.js?1628535635"></script>
    <script src="/scATACseq-workshop/js/hugo-learn.js?1628535635"></script>
    
        
            <script src="/scATACseq-workshop/mermaid/mermaid.js?1628535635"></script>
        
        <script>
            mermaid.initialize({ startOnLoad: true });
        </script>
    
    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-105947713-1', 'auto');
  ga('send', 'pageview');

</script>

  </body>
</html>

